fiona/ogrext.pyx

```diff
@@ -1367,14 +1597,15 @@ cdef class WritingSession(Session):
         None
 
         """
-        cdef void *cogr_driver
-        cdef void *cogr_feature
+        cdef OGRSFDriverH cogr_driver
+        cdef OGRFeatureH cogr_feature
         cdef int features_in_transaction = 0
-        cdef void *cogr_layer = self.cogr_layer
+        cdef OGRLayerH cogr_layer = self.cogr_layer
 
         if cogr_layer == NULL:
             raise ValueError("Null layer")
 
+        cdef OGRFeatureBuilder feat_builder = OGRFeatureBuilder(driver=collection.driver)
```

Comment:
Here's an optimization that I've overlooked for a while: reusing the feature builder.